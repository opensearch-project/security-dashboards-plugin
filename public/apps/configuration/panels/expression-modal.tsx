/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import {
  EuiLink,
  EuiCodeBlock,
  EuiModal,
  EuiModalBody,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
} from '@elastic/eui';

export function ExpressionModal(props: { title: string; expression: object }) {
  const [isModalVisible, setIsModalVisible] = React.useState(false);

  const closeModal = () => setIsModalVisible(false);

  const showModal = () => setIsModalVisible(true);

  let modal;

  if (isModalVisible) {
    modal = (
      <EuiOverlayMask>
        <EuiModal data-test-subj="expression-modal" onClose={closeModal}>
          <EuiModalHeader>
            <EuiModalHeaderTitle>{props.title}</EuiModalHeaderTitle>
          </EuiModalHeader>

          <EuiModalBody>
            <EuiCodeBlock fontSize="m" paddingSize="m" color="dark" overflowHeight={300} isCopyable>
              {JSON.stringify(props.expression, null, 2)}
            </EuiCodeBlock>
          </EuiModalBody>
        </EuiModal>
      </EuiOverlayMask>
    );
  }

  return (
    <div>
      <EuiLink data-test-subj="view-expression" onClick={showModal}>
        View expression
      </EuiLink>
      {modal}
    </div>
  );
}
