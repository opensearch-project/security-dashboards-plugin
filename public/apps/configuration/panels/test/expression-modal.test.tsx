/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { ExpressionModal } from '../expression-modal';

describe('Expression modal', () => {
  const setState = jest.fn();
  const useState = jest.spyOn(React, 'useState');
  const title = 'Title';
  const expression = { key: 'value1' };

  it('renders when isModalVisible = True', () => {
    useState.mockImplementationOnce(() => [true, setState]);
    const component = shallow(<ExpressionModal title={title} expression={expression} />);
    expect(component).toMatchSnapshot();
  });

  it('should set isModalVisible to True when click on View expression', () => {
    useState.mockImplementationOnce(() => [false, setState]);
    const component = shallow(<ExpressionModal title={title} expression={expression} />);
    component.find('[data-test-subj="view-expression"]').simulate('click');
    expect(setState).toHaveBeenCalledWith(true);
  });

  it('should set isModalVisible to false when close Expression modal', () => {
    useState.mockImplementationOnce(() => [true, setState]);
    const component = shallow(<ExpressionModal title={title} expression={expression} />);
    component.find('[data-test-subj="expression-modal"]').simulate('close');
    expect(setState).toHaveBeenCalledWith(false);
  });
});
