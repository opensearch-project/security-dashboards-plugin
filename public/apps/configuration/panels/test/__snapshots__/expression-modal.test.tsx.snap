// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression modal renders when isModalVisible = True 1`] = `
<div>
  <EuiLink
    data-test-subj="view-expression"
    onClick={[Function]}
  >
    View expression
  </EuiLink>
  <EuiOverlayMask>
    <EuiModal
      data-test-subj="expression-modal"
      onClose={[Function]}
    >
      <EuiModalHeader>
        <EuiModalHeaderTitle>
          <EuiText
            size="s"
          >
            <h2>
              Title
            </h2>
          </EuiText>
        </EuiModalHeaderTitle>
      </EuiModalHeader>
      <EuiModalBody>
        <EuiCodeBlock
          color="dark"
          fontSize="m"
          isCopyable={true}
          overflowHeight={300}
          paddingSize="m"
        >
          {
  "key": "value1"
}
        </EuiCodeBlock>
      </EuiModalBody>
    </EuiModal>
  </EuiOverlayMask>
</div>
`;
